File: app/src/main/java/com/gianessi/stargazers/network/NetworkManager.java
package com.gianessi.stargazers.network;

import com.gianessi.stargazers.models.Repo;
import com.gianessi.stargazers.models.User;
import com.gianessi.stargazers.models.UsersResponse;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkManager {

    public static final int FIRST_PAGE_INDEX = 1;
    private static final String BASE_URL = "https://api.github.com/";

    private static final NetworkManager ourInstance = new NetworkManager();

    public static NetworkManager getInstance() {
        return ourInstance;
    }

    private GitHubService service;

    private NetworkManager() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        service = retrofit.create(GitHubService.class);
    }

    public Call<UsersResponse> searchUsers(String query, int page){
        //Search query only in username field
        String encodedQuery = query;
        try{
            encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8.displayName());
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return service.searchUsers(String.format("%s+in:login", encodedQuery), page);
    }

    public Call<List<Repo>> listRepos(String username, int page){
        return service.listRepos(username, page);
    }

    public Call<List<User>> listStargazers(String username, String repoName, int page){
        return service.listStargazers(username, repoName, page);
    }

}

File: app/src/main/java/com/gianessi/stargazers/network/NetworkManager.kt
package com.gianessi.stargazers.network

import com.gianessi.stargazers.models.Repo
import com.gianessi.stargazers.models.User
import com.gianessi.stargazers.models.UsersResponse

import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

private const val BASE_URL = "https://api.github.com/"

class NetworkManager {

    private val service: GitHubService

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

        service = retrofit.create(GitHubService::class.java)
    }

    fun searchUsers(query: String, page: Int): Call<UsersResponse> {
        //Search query only in username field
        var encodedQuery = query
        try {
            encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8.displayName())
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }
        return service.searchUsers("$encodedQuery+in:login", page)
    }

    fun listRepos(username: String, page: Int): Call<List<Repo>> {
        return service.listRepos(username, page)
    }

    fun listStargazers(username: String, repoName: String, page: Int): Call<List<User>> {
        return service.listStargazers(username, repoName, page)
    }

    companion object {

        const val FIRST_PAGE_INDEX = 1

        val instance = NetworkManager()
    }

}
